src/core/provider.h

```diff
@@ -34,6 +34,37 @@
 namespace nvidia { namespace inferenceserver {
 
 class InferenceBackend;
+class SystemMemoryInferResponseProvider;
+
+//
+// Memory block used for accessing / storing data for providers
+// Note:
+//   - It is only responsible for deallocating memory obtained by ReserveBlock()
+//   - The functions are not thread-safe
+//
+class SystemMemoryBlock {
+ public:
+  // Create an SystemMemoryBlock object with 'byte_size'
+  SystemMemoryBlock(size_t byte_size);
+
+  // Create an SystemMemoryBlock that points to 'buffer' provided.
+  // 'buffer' is the memory that has been allocated. For instance,
+  // content returned by InferRequestProvider.GetNextInputContent().
+  // 'byte_size' is the size of the buffer
+  SystemMemoryBlock(const void* buffer, size_t byte_size);
+
+  // Return a pointer to the memory location.
+  const void* Buffer() const;
+
+  // Return the byte size of the memory
+  size_t ByteSize() const { return byte_size_; }
+
+ private:
+  friend class SystemMemoryInferResponseProvider;
```

Comment:
Why does it need to be  a friend. That implies something is wrong with the abstraction.
Just add MutableBuffer() for non-const buffer access.